src/util/exportToNDJson.js

```diff
@@ -88,37 +156,61 @@ const exportToNDJson = async (clientId, types, typeFilter, systemLevelExport, pa
 /**
  * Retrieves all documents from the requested collection and wraps them in an object with the collection name
  * @param {string} collectionName The collection of interest in the mongodb
- * @param {Object} typefilterLookup The entry in the _typeFilter that should be used to filter data to be exported
- * from the server
+ * @param {Object} searchParameterQueries The _typeFilter search parameter queries for the given resource type
+ * @param {Object} valueSetQueries list of ValueSet-related queries for the given resource type
  * @param {Array} patientIds Array of patient ids for which the returned documents should have references
  * @returns {Object} An object containing all data from the given collection name as well as the collection name
  */
-const getDocuments = async (collectionName, typefilterLookup, patientIds) => {
-  let query = {};
-  if (typefilterLookup[collectionName]) {
-    query = await processTypeFilter(typefilterLookup[collectionName]);
-  }
-  // Group export
+const getDocuments = async (collectionName, searchParameterQueries, valueSetQueries, patientIds) => {
+  let docs = [];
+  let patQuery = {};
+  let vsQuery = {};
+  // Create patient id query (for Group export only)
   if (patientIds) {
     if (patientIds.length == 0) {
       // if no patients in group, return no documents
       return { document: [], collectionName: collectionName };
     }
-    let patQuery;
     if (collectionName === 'Patient') {
       // simple patient id query
       patQuery = { id: { $in: patientIds } };
     } else {
       patQuery = await patientsQueryForType(patientIds, collectionName);
     }
-    // $and patient filtering with existing query
-    query = {
-      $and: [query, patQuery]
-    };
   }
 
-  const doc = await findResourcesWithQuery(query, collectionName, { projection: { _id: 0 } });
-  return { document: doc, collectionName: collectionName };
+  if (valueSetQueries) {
+    // extract codes from the ValueSet queries to use for filtering
+    vsQuery = await processVSTypeFilter(valueSetQueries);
+  }
+
+  if (searchParameterQueries) {
+    docs = searchParameterQueries.map(async q => {
+      let query = q;
+      // wherever we have a $match, we need to add another "and" operator containing the ValueSet and/or patient query
+      if (Object.keys(vsQuery).length > 0) {
+        query.filter(q => '$match' in q).forEach(q => (q['$match'] = { $and: [q['$match'], vsQuery] }));
+      }
+      if (Object.keys(patQuery).length > 0) {
+        query.filter(q => '$match' in q).forEach(q => (q['$match'] = { $and: [q['$match'], patQuery] }));
+      }
+      // grab the results from aggregation - has metadata about counts and data with resources in the first array position
+      const results = (await findResourcesWithAggregation(query, collectionName, { projection: { _id: 0 } }))[0];
```

Comment:
I think there might be something wonky going on here. I have a database populated with ~112 Encounters most have `status: 'finished'`, but some don't. When I do an export with `_typeFilter=Encounter?status=finished` I'm only getting 10 Encounters. I think mongo isn't returning a complete result or is paging the result on us.

Reply:
Ahh good catch Hoss! I also noticed that when you do this export, you only get 10 of the other resources, regardless of how many there actually are in the database. This doesn't happen with `_type` export, so maybe the solution can be found there.

Reply:
It looks like `fhirQueryBuilder` has a `resultsPerPage` param whose default value is 10. We could change that to be something larger, but what should that value be?

Reply:
I will need to look into this more, but I think we could keep the default value to 10 and iteratively write the results to the file (iterating over each page)? Since we want to capture all possible results

Reply:
With some help with Hoss, I think we can iterate over each page with the following:

![image](https://github.com/projecttacoma/bulk-export-server/assets/30158156/a31b9e40-db97-4fde-aff5-3df4eaee0b2b)

I have to do some more testing, but I believe this works!

Reply:
Tried this and the function ended up only going through the `while` loop once and still returning 10 results. I took a closer look at the actual query being used here, and `$facet.data` has a limit of 10, which I believe is the reason for still returning 10 results despite iterating over the cursor:
<img width="475" alt="Screenshot 2023-09-13 at 11 24 44 AM" src="https://github.com/projecttacoma/bulk-export-server/assets/87094479/c8f34b06-09ca-414c-9791-54d1a7191e0c">

Reply:
<img width="1053" alt="Screenshot 2023-09-13 at 12 24 35 PM" src="https://github.com/projecttacoma/bulk-export-server/assets/87094479/d9f0f88c-52f6-404a-b44e-1c2b2f3fc5e7">

Above is the `applyPaging` function in fhir-qb-mongo that gets used to build the query that is used for aggregation. To me, it seems like the logic for `numberOfPages` is sound, but the `$limit` may be restricting all the documents returned, which is causing only one page to exist? I will need to do some more investigating

Reply:
Ahhh I see! Sorry for the incorrect suggestion. I can also look into this 👍

Reply:
I don't think the suggestion is necessarily incorrect - I think it's good if we iterate over all the pages 👍 might just need some more investigation into `$limit` within the aggregation pipeline